utils/src/migration.rs

```diff
@@ -1,33 +1,215 @@
+        }
+        if s.contains("beta") {
+            let parts: Vec<_> = s[1..].split("beta").collect();
+            if parts.len() != 2 {
+                return Err(anyhow!("invalid api version format"));
+            }
+            let main_ver = parts[0].parse()?;
+            if parts[1].is_empty() {
+                return Ok(Self::Beta(main_ver, 0, PhantomData));
+            }
+            let sub_ver = parts[1].parse()?;
+            return Ok(Self::Beta(main_ver, sub_ver, PhantomData));
+        }
+        let main = s[1..].parse()?;
+        Ok(Self::Stable(main, PhantomData))
+    }
+}
+
+impl<CR> Display for ApiVersion<CR> {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let version = match *self {
+            ApiVersion::Alpha(main, sub, _) => {
+                if sub > 0 {
+                    format!("v{main}alpha{sub}")
+                } else {
+                    format!("v{main}alpha")
+                }
+            }
+            ApiVersion::Beta(main, sub, _) => {
+                if sub > 0 {
+                    format!("v{main}beta{sub}")
+                } else {
+                    format!("v{main}beta")
+                }
+            }
+            ApiVersion::Stable(main, _) => format!("v{main}"),
+        };
+        write!(f, "{version}")
+    }
+}
+
+impl<CR> PartialEq<Self> for ApiVersion<CR> {
+    #[inline]
+    #[allow(clippy::pattern_type_mismatch)]
+    fn eq(&self, other: &Self) -> bool {
+        let kind = matches!(
+            (self, other),
+            (Self::Alpha(_, _, _), Self::Alpha(_, _, _))
+                | (Self::Beta(_, _, _), Self::Beta(_, _, _))
+                | (Self::Stable(_, _), Self::Stable(_, _))
+        );
+        kind && self.main_version() == other.main_version()
+            && self.sub_version() == other.sub_version()
     }
-    if v2_parts > v1_parts {
-        return Ok(Ordering::Less);
+}
+
+impl<CR> PartialOrd for ApiVersion<CR> {
+    #[inline]
+    #[allow(clippy::pattern_type_mismatch)]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        // comparing order
+        // main version
+        // alpha < beta < stable
+        // sub version
+        if self.main_version() != other.main_version() {
+            return self.main_version().partial_cmp(&other.main_version());
+        }
+        match (self, other) {
+            (Self::Beta(_, sub, _), Self::Beta(_, other_sub, _))
+            | (Self::Alpha(_, sub, _), Self::Alpha(_, other_sub, _)) => sub.partial_cmp(other_sub),
+            (Self::Alpha(_, _, _) | Self::Beta(_, _, _), Self::Stable(_, _))
+            | (Self::Alpha(_, _, _), Self::Beta(_, _, _)) => Some(Ordering::Less),
+            (Self::Stable(_, _) | Self::Beta(_, _, _), Self::Alpha(_, _, _))
+            | (Self::Stable(_, _), Self::Beta(_, _, _)) => Some(Ordering::Greater),
+            _ => Some(Ordering::Equal),
```

Comment:
I suggest map the `Alpha` `Beta` `Stable` to number type for better readiblity, as the Eq and Ord of number is defined.

Reply:
done